File: src/nahyeon/BinarySearch/boj_1654_랜선자르기.java
package nahyeon.BinarySearch;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class boj_1654_랜선자르기 {

    static FastReader scan = new FastReader();
    static StringBuilder sb = new StringBuilder();

    static int K, N;
    static int[] A;

    static void input() {
        K = scan.nextInt();
        N = scan.nextInt();
        A = new int[K + 1];
        for (int i = 1; i <= K; i++) {
            A[i] = scan.nextInt();
        }
    }

    static boolean determination(long M) {
        // TODO: K 개의 일정하지 않은 길이의 랜선을 길이 M 으로 N 개를 만들 수 있는가
        long sum = 0;
        for (int i = 1; i <= K; i++) {
            sum += (A[i] / M);
        }
        return sum >= N;
    }

    static void pro() {

        Arrays.sort(A, 1, K + 1);

        // TODO: A[0...K] 범위에서 가능한 최대 길이 M 를 찾아낸다.
        long L = 1, R = Integer.MAX_VALUE, ans = 0;
        while (L <= R) {
            long mid = (L + R) / 2;
            if (determination(mid)) {
                ans = mid;
                L = mid + 1;
            } else {
                R = mid - 1;
            }
        }
        System.out.println(ans);
    }

    public static void main(String[] args) {
        input();
        pro();
    }

    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        String next() {
            while (st == null || !st.hasMoreTokens()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }
    }
}

File: src/hyewon/BinarySearch/boj_1654_랜선 자르기.kt
package hyewon.BinarySearch

import java.io.*
import java.util.*

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val bw = BufferedWriter(OutputStreamWriter(System.out))

    val line = StringTokenizer(br.readLine())
    val k = line.nextToken().toInt()
    val num = line.nextToken().toInt()

    val arr = Array(k){0L}
    for(i in 0 until k) {
        arr[i] = br.readLine().toLong()
    }

    var lo = 1L
    var hi = arr.max()
    var ans = 0L

    while(lo<=hi) {
        val mid = (lo+hi)/2L
        var cnt = 0L

        arr.forEach { cnt+=it/mid }

        if(cnt>=num) {
            ans = mid

            lo = mid + 1
        } else {
            hi = mid -1
        }
    }

    bw.write("$ans")
    bw.flush()
    bw.close()
}